include/revng/Pipeline/LLVMContainer.h

```diff
@@ -0,0 +1,201 @@
+    auto *ToMap = llvm::dyn_cast<llvm::MDNode>(Node);
+    if (ToMap == nullptr)
+      continue;
+
+    auto MayeMappedNode = Map.getMappedMD(ToMap);
+    if (not MayeMappedNode.hasValue())
+      continue;
+
+    auto *ToSet = llvm::dyn_cast<llvm::MDNode>(*MayeMappedNode);
+    if (ToSet == nullptr)
+      continue;
+    To.setMetadata(index, ToSet);
+  }
+}
+
+/// Implementation that can be derived by anyone so that multiple identical
+/// llvmcontainers can exists so that inspector kinds do not pollute each other.
+template<typename MostDerived>
+class LLVMContainerBase : public InspectableContainer<MostDerived> {
+private:
+  LLVMContainerImpl Container;
+
+public:
+  ~LLVMContainerBase() override {}
+
+  LLVMContainerBase(Context &Ctx, std::unique_ptr<llvm::Module> M) :
+    InspectableContainer<MostDerived>(Ctx), Container(std::move(M)) {}
+
+  const llvm::Module &getModule() const { return Container.getModule(); }
+  llvm::Module &getModule() { return Container.getModule(); }
+
+  void mergeBackImpl(MostDerived &&Other) final {
+    Container.mergeBackDerived(Other.Container);
+  }
+
+  llvm::Error storeToDisk(llvm::StringRef Path) const final {
+    return Container.storeToDisk(Path);
+  }
+
+  llvm::Error loadFromDisk(llvm::StringRef Path) final {
+    return Container.loadFromDisk(Path);
+  }
+
+  std::unique_ptr<ContainerBase>
+  cloneFiltered(const TargetsList &Targets) const final {
+    using InspectorT = LLVMContainerInspector<MostDerived>;
+    auto ToClone = InspectorT::enumerateAllFunctions(Targets, self());
+    auto ToClonedNotOwned = InspectorT::enumerateNonOnwedFunctions(self());
+
+    const auto Filter = [&ToClone, &ToClonedNotOwned](const auto &GlobalSym) {
+      if (not llvm::isa<llvm::Function>(GlobalSym))
+        return true;
+
+      const auto &F = llvm::cast<llvm::Function>(GlobalSym);
+      return ToClone.count(F) != 0 or ToClonedNotOwned.count(F) != 0;
+    };
+
+    llvm::ValueToValueMapTy Map;
+    auto Module = llvm::CloneModule(Container.getModule(), Map, Filter);
+
+    for (const llvm::Function &Original : Container.getModule().functions()) {
+      if (Original.isDeclaration())
+        cloneMetadata<llvm::Function>(Map, &Original);
+    }
+
+    return std::make_unique<MostDerived>(*this->Ctx, std::move(Module));
+  }
+
+  template<typename... LLVMPipePassess>
+  static PipeWrapper
+  wrapLLVMPassess(std::string LLVMModuleName, LLVMPipePassess &&...P) {
+    return PipeWrapper(LLVMPipeBase<MostDerived>(std::move(P)...),
+                       { std::move(LLVMModuleName) });
+  }
+
+  const LLVMContainerImpl &getContainer() const { return Container; }
+
+  LLVMContainerImpl &getContainer() { return Container; }
+
+  MostDerived &self() { return *static_cast<MostDerived *>(this); }
+
+  const MostDerived &self() const {
+    return *static_cast<const MostDerived *>(this);
+  }
+};
+
+/// regular llvm container used in most of the place
+class LLVMContainer : public LLVMContainerBase<LLVMContainer> {
```

Comment:
Can you hint at how a class that inherits from `LLVMContainerBase` but *it's not* `LLVMContainer` might look like?
I can't see it and I'd be up for squashing `LLVMContainerImpl`, `LLVMContainerBase` and `LLVMContainer` in a single class.

Reply:
This is needed because if someones needs to operate on a llvm container that is not poluted by the containers (either from the current project, upstream or downstream), it can create a new llvm container that is just like the base one but with a differen type and thus different inspectors.

Reply:
If we want to handle such situation, I don't think the right solution is a new type that is exactly 100% totally identical to the other one. This seems more like a run/load-time problems.

A better solution would be to have run-time "tags" into containers (for instance "PureLLVM", "Revng"). And then inspectors can decide to bind to a container only if it has a specific tag.
But I think this would be a new feature we can implement in the future if we need to.

Right now, by having `LLVMContainerBase<Class1>` and `LLVMContainerBase<Class2>`, you're doing something similar, basically using `::ID` as a tag, with the side-effect of duplicating all the code of the methods and pushing the rest of the codebase to have a template argument, which could be avoided.

Reply:
yes this is what is happening and this ensures us that if a llvm pipe is written for revng only it can be used only with the revng one rather than mistakenly be used on another one.

uhm, i guess i can use a compile time char* into the type to tell them apart, instead of just the type?

Reply:
> uhm, i guess i can use a compile time char* into the type to tell them apart, instead of just the type?

I'm not sure I understand, would this enable to squash together `LLVMContainerImpl`, `LLVMContainerBase` and `LLVMContainer`? If so, I'm happy!

> yes this is what is happening and this ensures us that if a llvm pipe is written for revng only it can be used only with the revng one rather than mistakenly be used on another one

If we want to support such a thing, I think we need to introduce a tag field that is passed in by the constructor and then inspectors/pipes can discriminate on them. If you can't come up with a simple solution for this now, let's just postpone and squash the 3 classes together.

Reply:
the 3 classes become one and what remains is like this

```cpp
template<const char** NameID>
  class LLVMContainerBase : public EnumeratableContainer<LLVMContainerBase<NameID>> {
    private:
     std::unique_ptr<llvm::Module> Module;
   
     using MyType = LLVMContainerBase<NameID>;
    
    public:
      ~LLVMContainerBase() override {}
      
      LLVMContainerBase(Context &Ctx,
                        std::unique_ptr<llvm::Module> M,
                        llvm::StringRef Name) : 
       EnumeratableContainer<MyType>(Ctx, Name), Module(std::move(M)) {}
...
};

extern const char* LLVMContainerName;                                                     ~                                                                                              
using LLVMContainer = LLVMContainerBase<&LLVMContainerName>; 
```

Reply:
i think this should be what you wanted

Reply:
Interesting solution!

I still think this could be a degree of freedom we can introduce if needed, but this makes me very happy!
I'd just rename `LLVMContainerName` to `LLVMContainerTypeID`, otherwise it could confuse people.